presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorBucketFunction.java

```diff
@@ -15,34 +15,91 @@
 
 import com.facebook.presto.spi.BucketFunction;
 import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
 import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
 import io.airlift.slice.XxHash64;
 
+import java.util.List;
+
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
 import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.Varchars.isVarcharType;
 import static com.google.common.base.Preconditions.checkArgument;
 
 public class RaptorBucketFunction
         implements BucketFunction
 {
-    private final int bucketCount;
+    private final HashFunction[] functions;
 
-    public RaptorBucketFunction(int bucketCount)
+    public RaptorBucketFunction(List<Type> types)
     {
-        checkArgument(bucketCount > 0, "bucketCount must be at least one");
-        this.bucketCount = bucketCount;
+        checkArgument(types != null, "types is is null");
+        checkArgument(!types.isEmpty(), "types is empty");
+
+        this.functions = new HashFunction[types.size()];
+        for (int i = 0; i < types.size(); i++) {
+            Type type = types.get(i);
+            if (type.equals(BIGINT)) {
+                functions[i] = bigintHashFunction();
+            }
+            else if (type.equals(INTEGER)) {
+                functions[i] = intHashFunction();
+            }
+            else if (isVarcharType(type)) {
+                functions[i] = varcharHashFunction();
+            }
+            else {
+                throw new PrestoException(NOT_SUPPORTED, "Bucketing not supported for type: " + type.getDisplayName());
+            }
+        }
     }
 
     @SuppressWarnings("NumericCastThatLosesPrecision")
     @Override
     public int getBucket(Page page, int position)
     {
         long hash = 0;
-        for (int i = 0; i < page.getChannelCount(); i++) {
+        int channelCount = page.getChannelCount();
+        for (int i = 0; i < channelCount; i++) {
             Block block = page.getBlock(i);
-            long value = BIGINT.getLong(block, position);
-            hash = (hash * 31) + XxHash64.hash(value);
+
+            HashFunction function = functions[i];
+            hash = (hash * 31) + function.hash(block, position);
         }
         int value = (int) (hash & Integer.MAX_VALUE);
-        return value % bucketCount;
+        return value % channelCount;
```

Comment:
This should be `bucketCount`. Otherwise, you only get as many buckets as the number of bucketing columns (so bucket on `[userid]` would produce only one bucket).